### DISCUSSION TITLE
Is anyone interested in generating a file of type hints for the luasnip snippet environment?

### DISCUSSION THREAD

**Discussion by JuanZoran** [L3MON4D3/LuaSnip#891] [Ideas]
![image](https://github.com/L3MON4D3/LuaSnip/assets/107862700/467d2af3-57c8-485b-8abc-985d94fb2518)

I am really grateful for this awesome snippet engine, it is so powerful that we can write almost any snippets that we need.
But as it is so powerful, and flexible so it is really a need to get type hints and docs from LSP, that is what I am trying to do.
**Just Like what `neodev.nvim` is doing**

But there are some problems I have to face.

- I am a newbie to Python so I can't handle the LuaSnip/DOC.md well, I have tried to use the `python.re` library to generate types and snippetNode hints for lua-language-server but failed ...
> You can check this meta file generated from LuaSnip DOC.md: https://github.com/JuanZoran/snippet_lua/blob/main/meta.lua

- I think the docs maybe need to be written with a defined format so that can be handled with a simple Python script or other tools. The best way I think is it should be a GitHub action...

if it is possible, it should provide multiple language support via GPT? I have tried to convert some meta files to the Chinese version
![image](https://github.com/L3MON4D3/LuaSnip/assets/107862700/492cf611-c7f8-49ac-a6d9-d9a73d68a11b)

I think there are two solutions for the current situation: 
1. To define a format for LuaSnip doc explain, then we can use Python or other tools to handle it.
2. Just use Copilot or GPT to try to generate this.

**Comment by JuanZoran** [comment]
Noticed that there already is a [PR](https://github.com/L3MON4D3/LuaSnip/pull/378) done that, but I don't know if it is impossible to implement...

**Comment by L3MON4D3** [comment]
First of all, getting the documentation directly in the editor seems like a pretty nice addition (although, I have to admit I'm more in the camp of "read documentation online/in a separate window", thus not completely enthusiastic about this😅)

I don't think we should rely on some LLM to generate the documentation, an error there would be pretty annoying.

I'd agree that 1) would be the easiest approach to this for now (it only requires a unified format for the source-files, and probably some tags to indicate where a function-definition starts), although generating `DOC.md`-entries from the source code (I'll just dub this approach 3) might be better overall since it would not require the step of generating a meta file for Luasnip (and, additionally, it seems nice to have the documentation right next to the code when changing it)
One thing I'm a bit worried about with 3) is that `DOC.md` might suffer a bit since cross-references might be harder to maintain/it would take longer to verify they actually work (have to first generate DOC.md)

After some more thought, 1) seems to be the best approach (I'd like to see `DOC.md` as good as possible, and this seems to be the only approach to guarantee that)

One concrete idea: in the meta-file, generate links to the corresponding section in `DOC.md`, for quick access to it (since there will always be things in that documentation not in the lsp-provided one, for example gifs)

My thoughts so far, wdyt?

**Reply by JuanZoran** [reply]
Actually, the lua-language-server support uses a `--doc` option to generate a markdown file from source code, I don't know whether you have known that...

You can find more info at [here](https://github.com/LuaLS/lua-language-server/wiki/Getting-Started)
![image](https://github.com/L3MON4D3/LuaSnip/assets/107862700/24321ea8-b779-48f2-a0a2-f16a975a35d0)


> like this : `lua-language-server --doc ...`

**Reply by L3MON4D3** [reply]
Oh I didn't know about that, pretty cool that it's built-in.
I don't think we can use it though, we have stuff like gifs, etc which probably can't be generated by it

**Reply by JuanZoran** [reply]
> Oh I didn't know about that, pretty cool that it's built-in. I don't think we can use it though, we have stuff like gifs, etc which probably can't be generated by it

I think maybe you can define a placeholder that tells the script to replace it with the file in the special folder.

**Reply by L3MON4D3** [reply]
But that script we'd have to write ourselves, right?
Couldn't find much info on `--doc`, but from what I found it does not seem to be customizable

Generating markdown for the function-definition, and the description of the various variables sounds like it might work, but some of these descriptions have links to other sections of `DOC.md`, and those would have to be maintained (so if one heading changes, one would have to search all of the source-files for references to it.. I don't really like that)

**Reply by JuanZoran** [reply]
Noticed that the lua-language-server comment supports using markdown syntax, i tried to paste the README.md link or image syntax to there and it words fine...
But I haven't tried to generate markdown file for that.. do u try that?

**Reply by L3MON4D3** [reply]
Oh right, of course..
I can definitely see the advantage of having the documentation right next to the code, but I fear that maintaining a well put-together DOC.md (which for me is definitely a priority) will be harder if it's content is spread over many different files.

But generating the function-documentation, and inserting it into DOC.md at the appropriate places might be a good idea, that way DOC.md can be more narrative, and contain intent and the like, and the source-files can just contain an accurate description of the function and its parameters, and that is appended to the corresponding section in DOC.md
